File: src/main/java/gargoyle/ct/prop/CTObservableProperty.java
package gargoyle.ct.prop;

import gargoyle.ct.pref.CTPropertyChangeListener;

import java.util.function.Function;

public interface CTObservableProperty<T> extends CTProperty<T> {
    void addPropertyChangeListener(CTPropertyChangeListener<T> listener);

    <T2> void bind(CTProperty<T2> property, Function<T, T2> mapper);

    void bind(CTProperty<T> property);

    <T2> void bindBi(CTObservableProperty<T2> property, Function<T, T2> mapper, Function<T2, T> mapper2);

    void bindBi(CTObservableProperty<T> property);

    void removePropertyChangeListener(CTPropertyChangeListener<T> listener);
}

File: util/src/main/kotlin/gargoyle/ct/util/prop/CTObservableProperty.kt
package gargoyle.ct.util.prop

import gargoyle.ct.util.pref.CTPropertyChangeListener
import java.util.function.Function

interface CTObservableProperty<T : Any> : CTProperty<T> {
    fun addPropertyChangeListener(listener: CTPropertyChangeListener<T>)
    fun <T2 : Any> bind(property: CTProperty<T2>, mapper: Function<T, T2>)
    fun bind(property: CTProperty<T>)
    fun <T2 : Any> bindBi(property: CTObservableProperty<T2>, mapper: Function<T, T2>, mapper2: Function<T2, T>)
    fun bindBi(property: CTObservableProperty<T>)
    fun removePropertyChangeListener(listener: CTPropertyChangeListener<T>)
}
